ulauncher/search/apps/AppStatDb.py

```diff
@@ -31,16 +28,5 @@ def get_most_frequent(self, limit=5):
         :rtype: class:`ResultList`
         """
 
-        # pylint: disable=relative-beyond-top-level
-        # import here to avoid circular deps.
-        from ulauncher.search.apps.AppResultItem import AppResultItem
-        from ulauncher.search.apps.AppDb import AppDb
-        app_db = AppDb.get_instance()
-
-        return [AppResultItem(i) for i in islice(
-            filter(None,
-                   map(lambda r: app_db.get_by_path(r[0]),
-                       sorted(self._records.items(),
-                              key=itemgetter(1),
-                              reverse=True))),
-            0, limit)]
+        top_results = sorted(self._records.items(), key=lambda rec: rec[1], reverse=True)[:limit]
```

Comment:
The previous code had a `filter`, so we don't think there is the possibility of an empty result any longer?

Reply:
I see that it was moved into the caller, and there it can use `AppResultItem` without an in-line import, which does feel a bit cleaner. The downside is that the limit is applied earlier now, so it is possible that we would have had valid (non-empty results), but the limit might actually elide all the "good" results, leaving only empty results that end up getting trimmed out by the caller. I could see especially during changing install/uninstall and upgrading of systems we could definitely have cases where the stuff in the "front" no longer exists. Then the `get_most_frequent(5)` might give none, or 1 or 2 entries that will actually results in a valid result item, even though we should be able to fill out all 5 if the limit happened later. Ultimately I'm not sure the code cleanliness is worth the UX (potential) impact.

Reply:
You are absolutely right about filtering being applied too early. :+1: I changed these methods so much back and forth that I confused myself in the process it seems.

The easiest way to fix this is to remove the limit argument and move [:limit] to the UlauncherWindow class as well, but I think I will move the entire method as a static method in AppResultItem instead.

Reply:
I moved AppStatDB into AppResultItem. `get_most_frequent` was the only public method used outside of AppResultItem.